### PR TITLE
Ben/docker compose initial changes

### PR DESCRIPTION
**Pull request by benzcash**
Some small tweaks to disconnect the services to enable containerization.

Currently requires https://github.com/zcash-hackworks/pyZcash/pull/1 and a `.zcash` and `.zcash-params` so we're not waiting for network sync on every change.

## Steps to test:
```
git clone git@github.com:benzcash/zfaucet.git test-zfaucet
cd test-zfaucet/
git checkout ben/docker-compose-initial-changes
git clone git@github.com:benzcash/pyZcash.git lib/pyZcash
cd pyZcash
git checkout ben/enable-rpc-params-in-env
cd ..
cp .env-example .env
<EDIT .env>
docker-compose up -d --build
```

For the `edit .env`, step most setting are self explanatory. 

The  `ZCASHD_DATADIR` and `ZCASHD_PARMDIR` should point to an existing zcashd location, make sure to stop your existing process!

Visit http://127.0.0.1:8080 to access the site.

It's possible the `zfaucet` container failed to start because `zcashd` isn't ready.

```
$ docker-compose ps
         Name                       Command                      State                    Ports          
---------------------------------------------------------------------------------------------------------
test-zfaucet_db_1        docker-entrypoint.sh postgres   Up                      0.0.0.0:5432->5432/tcp  
test-zfaucet_nginx_1     nginx -g daemon off;            Up                      0.0.0.0:8080->80/tcp    
test-zfaucet_zcashd_1    zcashd                          Up (health: starting)   0.0.0.0:18232->18232/tcp
test-zfaucet_zfaucet_1   docker/zfaucet/entrypoint.sh    Exit 1 
```

Just start it up

```
$ docker-compose up -d zfaucet
test-zfaucet_db_1 is up-to-date
test-zfaucet_zcashd_1 is up-to-date
Starting test-zfaucet_zfaucet_1 ... done
$ docker-compose ps
         Name                       Command                      State                    Ports          
---------------------------------------------------------------------------------------------------------
test-zfaucet_db_1        docker-entrypoint.sh postgres   Up                      0.0.0.0:5432->5432/tcp  
test-zfaucet_nginx_1     nginx -g daemon off;            Up                      0.0.0.0:8080->80/tcp    
test-zfaucet_zcashd_1    zcashd                          Up (health: starting)   0.0.0.0:18232->18232/tcp
test-zfaucet_zfaucet_1   docker/zfaucet/entrypoint.sh    Up                      0.0.0.0:8000->8000/tcp
```

Todo;
- healthcheck on containers
- fail2ban with Docker :thinking: 
- cert mapping for production

### CODE REVIEW

**Comment by charlieok** [comment]
What's the "fail2ban with Docker 🤔" item in the 'Todo' list about?

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      context: docker/zcashd
+    command: zcashd
+    ports:
+      - "18232:18232"
+    volumes:
+      - $ZCASHD_DATADIR:/home/zcashd/.zcash
+      - $ZCASHD_PARMDIR:/home/zcashd/.zcash-params
+
+volumes:
+  static_volume:
```
Comment (charlieok) [inline docker-compose.yml:55]: I'm curious what it means to have a key here (`static_volume`) with no value. What does this do?

File: docker/zfaucet/entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -e
+
+python manage.py makemigrations
+python manage.py migrate
+python manage.py collectstatic --no-input
+python manage.py healthcheck
+gunicorn --bind=0.0.0.0:8000 --workers=2 zfaucet.wsgi
```
Comment (mdr0id) [inline docker/zfaucet/entrypoint.sh]: minor nit: missing newline at EOF

File: docker/zcashd/zcash.conf [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+showmetrics=0
+logips=1
+testnet=1
+gen=1
+addnode=testnet.z.cash
+experimentalfeatures=1
+zshieldcoinbase=1
+maxconnections=1000
+rpcuser=zcashrpc
+rpcpassword=295c0dea2d014e0c29d140a077a86b3b
```
Comment (mdr0id) [inline docker/zcashd/zcash.conf]: minor nit: missing newline at EOF

File: docker/zcashd/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  && apt-get install -y zcash=$ZCASH_VERSION
+
+RUN adduser --quiet $ZCASHD_USER
+ADD zcash.conf /home/$ZCASHD_USER/.zcash/zcash.conf
+RUN chown -R $ZCASHD_USER:$ZCASHD_USER /home/$ZCASHD_USER
+USER zcashd
+WORKDIR /home/zcashd
+
+HEALTHCHECK --interval=5m --timeout=3s \
+    CMD zcash-cli getinfo || exit 1
```
Comment (mdr0id) [inline docker/zcashd/Dockerfile]: minor nit: missing newline EOF

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,9 @@
+FROM python:2
+ENV PYTHONUNBUFFERED 1
+
+WORKDIR /app
+ADD . /app
+RUN pip install -r requirements.txt && \
+    pip install -r requirements-prod.txt  
+
+ENTRYPOINT ["docker/zfaucet/entrypoint.sh"]
```
Comment (mdr0id) [inline Dockerfile]: minor nit: missing newline on EOF

**Comment by benzcash** [comment]
> What's the "fail2ban with Docker " item in the 'Todo' list about?

Fail2ban works on iptables, which the containerized apps don't have access to.

So, either we figure out a way to get them access or an find alternative anti-abuse method. I'm going to look at some alternatives.

File: docker-compose.yml (hunk 1 above)
Comment (benzcash) [reply docker-compose.yml:55]: https://docs.docker.com/compose/compose-file/#volume-configuration-reference

Long story short, this is a way to attach a volume to the containers that isn't an existing host file path. It's "private" to the containers (but obviously still accessible if you really want to). Declaring the keyword makes it available for use in the service definitions.

In the instance, it's so django can "collect" the static files to a location that the nginx container can access for serving.

We certainly can use an existing host path, but I personally like it being self contained.

**Comment by mdr0id** [comment]
I keep getting errors trying to build the image:

```
 ---> Running in 3fdf88a382a8
Warning: apt-key output should not be parsed (stdout is not a terminal)
Executing: /tmp/apt-key-gpghome.sSJKMwXCQI/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 6DEF3BAF272766C0
gpg: keyserver receive failed: Server indicated a failure
ERROR: Service 'zcashd' failed to build: The command '/bin/sh -c apt-key adv --keyserver keyserver.ubuntu.com --recv-keys $ZCASH_SIGNING_KEY_ID   && echo "deb [arch=amd64] https://apt.z.cash/ jessie main" > /etc/apt/sources.list.d/zcash.list   && apt-get update   && apt-get install -y zcash=$ZCASH_VERSION' returned a non-zero code: 2
```

**Review by mdr0id — COMMENTED** [review]
A few minor nits.

I aslo keep getting an error during the image build through docker-compose.

File: docker/zcashd/Dockerfile (hunk 4 above)
Comment (benzcash) [reply docker/zcashd/Dockerfile]: Updated!

File: docker/zfaucet/entrypoint.sh (hunk 2 above)
Comment (benzcash) [reply docker/zfaucet/entrypoint.sh]: Updated!

File: Dockerfile (hunk 5 above)
Comment (benzcash) [reply Dockerfile]: Updated!

**Comment by benzcash** [comment]
> I keep getting errors trying to build the image:
> `.env`
> 
> ```
> POSTGRES_PASSWORD=notsecure
> POSTGRES_HOST=db
> ENVIRONMENT=dev
> RPCUSER=zcashrpc
> RPCPASSWORD=notsecure
> ZCASHD_DATADIR=~/.zcash/testnet3/
> ZCASHD_PARMDIR=~/.zcash-params/
> ```
> 
> ```
>  ---> Running in 3fdf88a382a8
> Warning: apt-key output should not be parsed (stdout is not a terminal)
> Executing: /tmp/apt-key-gpghome.sSJKMwXCQI/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys 6DEF3BAF272766C0
> gpg: keyserver receive failed: Server indicated a failure
> ERROR: Service 'zcashd' failed to build: The command '/bin/sh -c apt-key adv --keyserver keyserver.ubuntu.com --recv-keys $ZCASH_SIGNING_KEY_ID   && echo "deb [arch=amd64] https://apt.z.cash/ jessie main" > /etc/apt/sources.list.d/zcash.list   && apt-get update   && apt-get install -y zcash=$ZCASH_VERSION' returned a non-zero code: 2
> ```

Updated the Dockerfile to force port 80 for the apt-key download.

@mdr0id can you please try to build again and see if that resolves your issue?

**Comment by mdr0id** [comment]
I was able to get past my prior mentioned issue. Now I am getting 502 on 127.0.0.1:8080

```
File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='zcashd', port=18232): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f3155826250>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

**Comment by torizoemitchell** [comment]
When I worked though this with Ben, the only issue we had was with the "race condition" between zcashd and the zfaucet, which Ben said he was aware of and was working on.

**Comment by mdr0id** [comment]
We should remove commit 5910dee6484417a466442092d96e678b2cf5f92b and keep it at 1.0 payout

File: faucet/templates/faucet/modal.html [hunk 6]
```diff
@@ -10,11 +10,15 @@ <h5 class="modal-title">Help fund the faucet!</h5>
       <div class="modal-body">
         <p >
           Transparent donation address:</br>
-          <span class="taddress">tmKBPqa8qqKA7vrGq1AaXHSAr9vqa3GczzK</span>
+          <span class="taddress">tmP5cmDU66W2puHFoqfu8Zqe6M4mpsDk5fS</span>
```
Comment (benzcash) [inline faucet/templates/faucet/modal.html]: @charlieok do you know what addresses we should use for donations to the faucet?

File: faucet/templates/faucet/modal.html [hunk 7]
```diff
@@ -10,11 +10,15 @@ <h5 class="modal-title">Help fund the faucet!</h5>
...
         </p>
         <p>
-          Shielded donation address:</br>
-          <span class="zaddress">ztbx5DLDxa5ZLFTchHhoPNkKs57QzSyib6UqXpEdy76T1aUdFxJt1w9318Z8DJ73XzbnWHKEZP9Yjg712N5kMmP4QzS9iC9</span>
+          Sprout Shielded donation address:</br>
+          <span class="zaddress">ztSwdDwPhpUZ447YU1BqjxrvutHfu2AyENwUohhTMhnWHreAEHTELhRLvqkARmCSudW1GAcrg58TVaqT7oTH1ohFA7k7V11</span>
+        </p>
+        <p>
+          Sapling Shielded donation address:</br>
+          <span class="zaddress">ztestsapling1603ydy9hg79lv5sv9pm5hn95cngfv4qpd6y54a8wkyejn72jl30a4pfhw8u00p93mu4nj6qxsqg</span>
```
Comment (benzcash) [inline faucet/templates/faucet/modal.html]: @charlieok Same with the sapling addresses

File: docker/zfaucet/entrypoint.sh [hunk 8]
```diff
@@ -5,6 +5,7 @@ set -e
 python manage.py makemigrations
 python manage.py migrate
 python manage.py collectstatic --no-input
-python manage.py check_zcashd --wait 20
+python manage.py check_zcashd --wait 120
```
Comment (mdr0id) [inline docker/zfaucet/entrypoint.sh:8]: This is a better interval, especially when block dimes decrease with blossum activation

**Review by mdr0id — APPROVED** [review]
utACK

File: faucet/templates/faucet/modal.html (hunk 6 above)
Comment (charlieok) [reply faucet/templates/faucet/modal.html]: I am not sure... is this question still unresolved?

File: faucet/templates/faucet/modal.html (hunk 7 above)
Comment (charlieok) [reply faucet/templates/faucet/modal.html]: is this question still unresolved?

Comment (benzcash) [reply faucet/templates/faucet/modal.html]: Kind of, updated it again so it'll use the existing faucet.testnet.z.cash.

If there is some other, general addresses, we could use those. But I think this is fine for now.

**Review by charlieok — APPROVED** [review]
ACK

**Pull request merged**